draft-ietf-tls-dtls-connection-id.md

```diff
@@ -140,6 +140,14 @@ indicates that the server will send with the client's CID but does not
 wish the client to use a CID (or again, alternately, to use a
 zero-length CID).
 
+Whenever a zero-length CID has been negotiated then the RFC 6347-defined 
```

Comment:
I think that you need to explicitly mention the MAC calculation in this paragraph.  A zero-length CID will naturally look the same on the wire, it's how the MAC is calculated is the real trap here.

Reply:
> A zero-length CID will naturally look the same on the wire

I didn't get it. A zero-length CID uses the RFC6347 record layer and so the types used are defined in https://tools.ietf.org/html/rfc5246#section-6.2.1 

```
enum {
change_cipher_spec(20), alert(21), handshake(22),
application_data(23), (255)
} ContentType;
```

if the cid is none-empty, the `type` is the new one `tls12_cid(25)`.

So I'm not sure, do you have a use-case, which makes a difference using a empty cid or no cid? I can't see the benefit in distinguish that.

Reply:
This looks like a good way forward,  can we clarify this in the PR?

Reply:
Would it be possible, that the use of "empty CIDs" could be addressed explicit?

My impression is, before introducing the `uint8 zeros[length_of_padding] ` an empty CID in the extension was used by a client to indicate, that the client will be able to accept a CID from the server. It was not to force the server to use the new record layer when sending data to the client even with a empty CID.

With the introduction of the new record layer with the `uint8 zeros[length_of_padding] `, this draft has a additional functionality, it could be used to obfuscate the payload length. So, if this extension is not only intended to solve the address (NAT) issue, and it is intended to offer such a "payload length  obfuscation", then that should be made more explicit.

Reply:
> If the CID presence/absence doesn't match what was negotiated, that's a MAC error

I'm not sure about that. If in "Absence" a RFC6347 record is used, and it matches the right connection with the address:port, this will not cause a MAC error (it's a valid RFC6347 MAC). instead, if the "cid requires cid" (see #28) should be implemented and a packet with a record without CID MUST be dropped, that will be an (easy) additional check.

Reply:
> If the CID presence/absence doesn't match what was negotiated, that's a MAC error

FMPOV, that would only be true, if https://github.com/tlswg/dtls-conn-id/blob/master/draft-ietf-tls-dtls-connection-id.md#record-payload-protection applies for **all** records, not only tls_cid.
 
> Moreover, authenticating the whole header is just good practice.

If the new MAC is applied independent from the record type, the cid-values are not from the header.

But I would prefer a interpretation, where the new MAC is bound to the new record type.

Reply:
> If the CID presence/absence doesn't match what was negotiated, that's a MAC error

Or is the intention to **assume**, that's a MAC error (and be processed as such a MAC error). But it's not an error caused by the calculation of the MAC, because the MAC of a RFC6347 record would be valid for itself.